file-3.37/file.man
.TH FILE __CSECTION__ "October 2001" "Debian/GNU Linux" "Copyrighted but distributable"
.\" $Id: file.man,v 1.39 2001/04/27 22:48:33 christos Exp $
.SH NAME
file
\- determine file type
.SH SYNOPSIS
.B file
[
.B \-bciknsvzL
]
[
.B \-f
.I namefile
]
[
.B \-m 
.I magicfiles
]
.I file
\&...
.br
.B file
.B -C
[
.B \-m 
magicfile ]
.SH DESCRIPTION
This manual page documents version __VERSION__ of the
.B file
command.
.PP
.B File
tests each argument in an attempt to classify it.
There are three sets of tests, performed in this order:
filesystem tests, magic number tests, and language tests.
The
.I first
test that succeeds causes the file type to be printed.
.PP
The type printed will usually contain one of the words
.B text
(the file contains only
printing characters and a few common control
characters and is probably safe to read on an
.SM ASCII
terminal),
.B executable
(the file contains the result of compiling a program
in a form understandable to some \s-1UNIX\s0 kernel or another),
or
.B data
meaning anything else (data is usually `binary' or non-printable).
Exceptions are well-known file formats (core files, tar archives)
that are known to contain binary data.
When adding local definitions to
.IR /etc/magic ,
.BR "preserve these keywords" .
People depend on knowing that all the readable files in a directory
have the word ``text'' printed.
Don't do as Berkeley did and change ``shell commands text''
to ``shell script''.
Note that the file
.I __MAGIC__
is built mechanically from a large number of small files in
the subdirectory
.I Magdir
in the source distribution of this program.
.PP
The filesystem tests are based on examining the return from a
.BR stat (2)
system call.
The program checks to see if the file is empty,
or if it's some sort of special file.
Any known file types appropriate to the system you are running on
(sockets, symbolic links, or named pipes (FIFOs) on those systems that
implement them)
are intuited if they are defined in
the system header file
.IR <sys/stat.h>  .
.PP
The magic number tests are used to check for files with data in
particular fixed formats.
The canonical example of this is a binary executable (compiled program)
.I a.out
file, whose format is defined in 
.I a.out.h
and possibly
.I exec.h
in the standard include directory.
These files have a `magic number' stored in a particular place
near the beginning of the file that tells the \s-1UNIX\s0 operating system
that the file is a binary executable, and which of several types thereof.
The concept of `magic number' has been applied by extension to data files.
Any file with some invariant identifier at a small fixed
offset into the file can usually be described in this way.
The information identifying these files is read from
.I /etc/magic
and the compiled
magic file
.I __MAGIC__.mgc ,
or 
.I __MAGIC__
if the compile file does not exist.
.PP
If a file does not match any of the entries in the magic file,
it is examined to see if it seems to be a text file.
ASCII, ISO-8859-x, non-ISO 8-bit extended-ASCII character sets
(such as those used on Macintosh and IBM PC systems),
UTF-8-encoded Unicode, UTF-16-encoded Unicode, and EBCDIC
character sets can be distinguished by the different
ranges and sequences of bytes that constitute printable text
in each set.
If a file passes any of these tests, its character set is reported.
ASCII, ISO-8859-x, UTF-8, and extended-ASCII files are identified
as ``text'' because they will be mostly readable on nearly any terminal;
UTF-16 and EBCDIC are only ``character data'' because, while
they contain text, it is text that will require translation
before it can be read.
In addition,
.B file
will attempt to determine other characteristics of text-type files.
If the lines of a file are terminated by CR, CRLF, or NEL, instead
of the Unix-standard LF, this will be reported.
Files that contain embedded escape sequences or overstriking
will also be identified.
.PP
Once
.B file
has determined the character set used in a text-type file,
it will
attempt to determine in what language the file is written.
The language tests look for particular strings (cf
.IR names.h )
that can appear anywhere in the first few blocks of a file.
For example, the keyword
.B .br
indicates that the file is most likely a
.BR troff (1)
input file, just as the keyword 
.B struct
indicates a C program.
These tests are less reliable than the previous
two groups, so they are performed last.
The language test routines also test for some miscellany
(such as 
.BR tar (1)
archives).
.PP
Any file that cannot be identified as having been written
in any of the character sets listed above is simply said to be ``data''.
.SH OPTIONS
.TP 8
.I "\-b, \-\-brief"
Do not prepend filenames to output lines.
.TP 8
.I "\-c, \-\-checking\-printout"
Cause a checking printout of the parsed form of the magic file.
This is usually used in conjunction with 
.B \-m
to debug a new magic file before installing it.
.TP 8
.I "\-C, \-\-compile"
Write a magic.mgc output file that contains a pre-parsed version of
file.
.TP 8
.I "\-f, \-\-files-from namefile"
Read the names of the files to be examined from 
.I namefile
(one per line) 
before the argument list.
Either 
.I namefile
or at least one filename argument must be present;
to test the standard input, use ``\-'' as a filename argument.
.TP
.I "\-i, --mime"
Causes the file command to output mime type strings rather than the more
traditional human readable ones. Thus it may say
``text/plain; charset=us-ascii''
rather
than ``ASCII text''. In order for this option to work, file changes the way
it handles files recognised by the command itself (such as many of the
text file types, directories etc), and makes use of an alternative
``magic'' file.
(See ``FILES'' section, below).
.TP 8
.I "\-k, --keep-going"
Don't stop at the first match, keep going.
.TP 8
.I "\-m, \-\-magic-file list"
Specify an alternate list of files containing magic numbers.
This can be a single file, or a colon-separated list of files.
.TP 8
.I "\-n, \-\-no-buffer"
Force stdout to be flushed after checking each file. This is only useful if
checking a list of files. It is intended to be used by programs that want
filetype output from a pipe.
.TP 8
.B \-v
Print the version of the program and exit.
.TP 8
.I "\-z, \-\-uncompress"
Try to look inside compressed files.
.TP 8
.I "\-L, \-\-dereference"
This
option causes symlinks to be followed, as the like-named option in
.BR ls (1).
(on systems that support symbolic links).
.TP 8
.I "\-s, \-\-special-files"
Normally,
.B file
only attempts to read and determine the type of argument files which
.BR stat (2)
reports are ordinary files.
This prevents problems, because reading special files may have peculiar
consequences.
Specifying the
.BR \-s
option causes
.B file
to also read argument files which are block or character special files.
This is useful for determining the filesystem types of the data in raw
disk partitions, which are block special files.
This option also causes
.B file
to disregard the file size as reported by
.BR stat (2)
since on some systems it reports a zero size for raw disk partitions.
.TP
.I "\-\-help"
Print a help message and exit.
.TP
.I "\-\-version"
Print version information and exit.
.SH FILES
.TP
.I __MAGIC__.mgc
Default compiled list of magic numbers
.TP
.I __MAGIC__
Default list of magic numbers
.TP
.I __MAGIC__.mime
Default list of magic numbers, used to output mime types when the -i option
is specified.
.TP
.I /etc/magic
Local additions to magic wisdom.

.SH ENVIRONMENT
The environment variable
.B MAGIC
can be used to set the default magic number files.
.SH SEE ALSO
.BR magic (__FSECTION__)
\- description of magic file format.
.br
.BR strings (1), " od" (1), " hexdump(1)"
\- tools for examining non-textfiles.
.SH STANDARDS CONFORMANCE
This program is believed to exceed the System V Interface Definition
of FILE(CMD), as near as one can determine from the vague language
contained therein. 
Its behaviour is mostly compatible with the System V program of the same name.
This version knows more magic, however, so it will produce
different (albeit more accurate) output in many cases. 
.PP
The one significant difference 
between this version and System V
is that this version treats any white space
as a delimiter, so that spaces in pattern strings must be escaped.
For example,
.br
>10	string	language impress\ 	(imPRESS data)
.br
in an existing magic file would have to be changed to
.br
>10	string	language\e impress	(imPRESS data)
.br
In addition, in this version, if a pattern string contains a backslash,
it must be escaped.  For example
.br
0	string		\ebegindata	Andrew Toolkit document
.br
in an existing magic file would have to be changed to
.br
0	string		\e\ebegindata	Andrew Toolkit document
.br
.PP
SunOS releases 3.2 and later from Sun Microsystems include a
.BR file (1)
command derived from the System V one, but with some extensions.
My version differs from Sun's only in minor ways.
It includes the extension of the `&' operator, used as,
for example,
.br
>16	long&0x7fffffff	>0		not stripped
.SH MAGIC DIRECTORY
The magic file entries have been collected from various sources,
mainly USENET, and contributed by various authors.
Christos Zoulas (address below) will collect additional
or corrected magic file entries.
A consolidation of magic file entries 
will be distributed periodically.
.PP
The order of entries in the magic file is significant.
Depending on what system you are using, the order that
they are put together may be incorrect.
.SH EXAMPLES
.nf
$ file file.c file /dev/hda
file.c:   C program text
file:     ELF 32-bit LSB executable, Intel 80386, version 1,
          dynamically linked, not stripped
/dev/hda: block special

$ file -s /dev/hda{,1,2,3,4,5,6,7,8,9,10}
/dev/hda:   x86 boot sector
/dev/hda1:  Linux/i386 ext2 filesystem
/dev/hda2:  x86 boot sector
/dev/hda3:  x86 boot sector, extended partition table
/dev/hda4:  Linux/i386 ext2 filesystem
/dev/hda5:  Linux/i386 swap file
/dev/hda6:  Linux/i386 swap file
/dev/hda7:  Linux/i386 swap file
/dev/hda8:  Linux/i386 swap file
/dev/hda9:  empty
/dev/hda10: empty

$ file -i file.c file /dev/hda
file.c:      text/x-c
file:        application/x-executable, dynamically linked (uses shared libs), not stripped
/dev/hda:    application/x-not-regular-file

.fi
.SH HISTORY
There has been a 
.B file
command in every \s-1UNIX\s0 since at least Research Version 6
(man page dated January 16, 1975).
The System V version introduced one significant major change:
the external list of magic number types.
This slowed the program down slightly but made it a lot more flexible.
.PP
This program, based on the System V version,
was written by Ian Darwin <ian@darwinsys.com>
without looking at anybody else's source code.
.PP
John Gilmore revised the code extensively, making it better than
the first version.
Geoff Collyer found several inadequacies
and provided some magic file entries.
Contributions by the `&' operator by Rob McMahon, cudcv@warwick.ac.uk, 1989.
.PP
Guy Harris, guy@netapp.com, made many changes from 1993 to the present.
.PP
Primary development and maintenance from 1990 to the present by
Christos Zoulas (christos@astron.com).
.PP
Altered by Chris Lowth, chris@lowth.com, 2000:
Handle the ``-i'' option to output mime type strings and using an alternative
magic file and internal logic.
.PP
Altered by Eric Fischer (enf@pobox.com), July, 2000,
to identify character codes and attempt to identify the languages
of non-ASCII files.
.PP
The list of contributors to the "Magdir" directory (source for the
/etc/magic
file) is too long to include here. You know who you are; thank you.
.SH LEGAL NOTICE
Copyright (c) Ian F. Darwin, Toronto, Canada, 1986-1999.
Covered by the standard Berkeley Software Distribution copyright; see the file
LEGAL.NOTICE in the source distribution.
.PP
The files
.I tar.h
and
.I is_tar.c
were written by John Gilmore from his public-domain
.B tar
program, and are not covered by the above license.
.SH BUGS
There must be a better way to automate the construction of the Magic
file from all the glop in Magdir. What is it?
Better yet, the magic file should be compiled into binary (say,
.BR ndbm (3)
or, better yet, fixed-length
.SM ASCII
strings for use in heterogenous network environments) for faster startup.
Then the program would run as fast as the Version 7 program of the same name,
with the flexibility of the System V version.
.PP
.B File
uses several algorithms that favor speed over accuracy,
thus it can be misled about the contents of
text
files.
.PP
The support for
text
files (primarily for programming languages)
is simplistic, inefficient and requires recompilation to update.
.PP
There should be an ``else'' clause to follow a series of continuation lines.
.PP
The magic file and keywords should have regular expression support.
Their use of
.SM "ASCII TAB"
as a field delimiter is ugly and makes
it hard to edit the files, but is entrenched.
.PP
It might be advisable to allow upper-case letters in keywords
for e.g.,
.BR troff (1)
commands vs man page macros.
Regular expression support would make this easy.
.PP
The program doesn't grok \s-2FORTRAN\s0.
It should be able to figure \s-2FORTRAN\s0 by seeing some keywords which 
appear indented at the start of line.
Regular expression support would make this easy.
.PP
The list of keywords in 
.I ascmagic
probably belongs in the Magic file.
This could be done by using some keyword like `*' for the offset value.
.PP
Another optimisation would be to sort
the magic file so that we can just run down all the
tests for the first byte, first word, first long, etc, once we
have fetched it.  Complain about conflicts in the magic file entries.
Make a rule that the magic entries sort based on file offset rather
than position within the magic file?
.PP
The program should provide a way to give an estimate 
of ``how good'' a guess is.
We end up removing guesses (e.g. ``From '' as first 5 chars of file) because
they are not as good as other guesses (e.g. ``Newsgroups:'' versus
``Return-Path:'').  Still, if the others don't pan out, it should be
possible to use the first guess.  
.PP
This program is slower than some vendors' file commands.
The new support for multiple character codes makes it even slower.
.PP
This manual page, and particularly this section, is too long.
.SH AVAILABILITY
You can obtain the original author's latest version by anonymous FTP
on
.B ftp.astron.com
in the directory
.I /pub/file/file-X.YY.tar.gz
.PP
This
.B Debian
version adds long options and corrects some bugs. It can be
obtained from every site carrying a
.B Debian
distribution (ftp.debian.org and mirrors).


proftpd/NEWS
$Id: NEWS,v 1.594 2004/09/04 18:50:03 castaglia Exp $

1.2.10 - Released 04-Sep-2004
--------------------------------
- Bug 2440 - Unable to use PAM authentication properly.  Use a "*" after
  the module name in an AuthOrder directive to indicate that an auth
  module is authoritative.
- Bug 2441 - AIX5 portability bug with mod_auth_unix, mod_auth_file.
- Bug 2442 - Segfault in FreeBSD PAM library with long login names.
- Bug 2445 - AuthUserFile in <Global> context overrides <VirtualHost> setting.
- Bug 2444 - Use of sendfile() does not interoperate well with RFC2228
  security mechanisms.  Using sendfile(2) to send data bypasses the handling
  of the data by RFC2228 security mechanisms (such as those provided by
  mod_tls).  So if security mechanisms are detected, do not use sendfile().
- Scrub the ScoreboardFile for stale sessions in inetd mode.
- Bug 2427 - proftpd gets a memory fault when run from ssh batch mode.

1.2.10rc3 - Released 13-Jul-2004
--------------------------------
- Fixed typo that prevented 1.2.10rc2 from compiling.

1.2.10rc2 - Released 13-Jul-2004
---------------------------------
- Bug 2396 - NLST command doesn't understand options.  This was caused by
  the solution for Bug 2322.  However, it is not a popular solution, so
  NLST will once again handle options, but only the relevant options.
- Bug 2034 - Add support for a "graceful shutdown" signal. See
  contrib/mod_ctrls_admin.html#shutdown for details.
- Bug 2400 - <Class> search order is wrong.  The documentation correctly
  stated that <Class> sections are matched in order of definition, but the
  code has the match order in the reverse order of definition.
- Bug 2401 - MaxClientsPerClass only checks first directive in config file.
- Bug 2399 - Rename start/stop control actions to up/down.
- Bug 2082 - Add mod_rewrite "replaceall" builtin function.  See the
  RewriteMap documentation for more details.
- Bug 2403 - Sending SIGHUP to proftpd stops it when using Classes.  The fix
  for Bug #2400 could result in an infinite loop during a SIGHUP.
- Bug 2405 - "LIST *" should not list dotfiles.
- Bug 2366 - Add support for -h list option.
- Bug 2332 - SO_OOBINLINE error after upgrading proftpd from 1.2.6 to 1.2.9.
  This is due mostly to a change in the logging; a check for error values
  and logging of them was added.  The setting of this particular socket option
  has been moved earlier in the session, as it was found that short-lived
  TCP connections, as from monitoring systems, would cause this error.
- Bug 2407 - mod_auth_file does not allow for proper cascading of "end" and
  "set" auth requests.
- Bug 2410 - CreateHome always copies skel directory.
- Bug 2336 - Use of /dev/log on Solaris leads to kernel memory leak.
  ProFTPD's use of the /dev/log device on Solaris was tickling a Solaris
  kernel bug that caused the Solaris kernel to leak memory.
- Added a TimeoutLinger directive to complement the --enable-timeout-linger
  configure option.
- Bug 2125 - -vv command line switch should list versions of modules.
- Bug 2420 - Name field is not escaped before querying database.
  The mod_quotatab module was not properly escaping the name string it
  used when looking up records from SQL databases.
- Bug 2424 - SQLDefaultHomedir overrides column value.
- Bug 2411 - Caching effects cause RNTO to fail if AllowOverwrite is off and
  target path does not exist.
- Bug 2422 - %v not working in SQLNamedQuery.
- Bug 2418 - chmod returns 550 with filename containing multiple spaces.
- Bug 2431 - mod_sql does not use UID/GID properly in cache lookups.
- Bug 2303 - Problem evaluating multiple <Class> rules.
- Bug 2419 - Ability to disable TLSRequired on per-user basis (e.g. for
  anonymous logins).
- Bug 2438 - Display variable %z not expanded properly.
- Bug 2439 - <Limit CWD> doesn't work.

1.2.10rc1 - Released 28-Apr-2004
---------------------------------
- Bug 2135 - Add ability to handle passphrase-protected server keys.  mod_tls
  can now properly prompt for passphrases for protected server certificate
  keys when the daemon is starting up.
- Bug 2086 - Add limits for PORT, PASV.  This means that now one can use
  <Limit> to place access controls on the PORT and PASV commands.  This
  applies to the EPRT and EPSV commands as well.
- Bug 2174 - mod_auth_unix should not act authoritatively.  This was causing
  problems when using mod_auth_unix.c and the AuthOrder configuration
  directive.
- Bug 2098 - Added SetEnv and UnsetEnv configuration directives.
- Bug 2271 - Improper autoconf check for getaddrinfo() on Tru64 UNIX 5.1.
  The getaddrinfo symbol is a macro, not a function, on that platform.
- Bug 2255 - RADIUS Service-Type should reflect attribute expectations.
- Added Event API.
- Bug 2272 - Address/port collision check needs to handle DefaultAddress.
- Bug 2072 - Add Controls API.
  This API includes a new program, ftpdctl, that is used to communicate
  directly with the proftpd daemon via a new core module, mod_ctrls.  For
  this new functionality to be used, proftpd must be configured using the
  added --enable-ctrls option.
- Bug 2015 - Add AND, OR keywords to Allow/DenyUser directives.
  The AllowUser, DenyUser, AllowGroup, and DenyGroup directives now take
  an optional keyword that indicates what type of expression they are:
  AND, OR, or regex.  By default, AllowUser and DenyUser are OR expressions,
  and AllowGroup and DenyGroup are AND expressions.  For example:

    AllowUser regex ^ftp
    DenyUser AND dave,bob
    AllowGroup OR web,doc

  These demonstrate that the optional keyword modifier must be the first
  parameter in the configuration directive.
- Bug 2046 - Change RFNR and RNTO logging class to WRITE.  This means that
  ExtendedLogs that use the WRITE logging class will now include the
  RNFR and RNTO commands.
- Mac OS X 10.3 portability fixes.
- Bug 2274 - Default server only binds to one IP address of host if
  the --enable-ipv6 configure option is used.  ProFTPD will now properly

file-3.37/file.c
/*
 * file - find type of a file or files - main program.
 *
 * Copyright (c) Ian F. Darwin, 1987.
 * Written by Ian F. Darwin.
 *
 * This software is not subject to any license of the American Telephone
 * and Telegraph Company or of the Regents of the University of California.
 *
 * Permission is granted to anyone to use this software for any purpose on
 * any computer system, and to alter it and redistribute it freely, subject
 * to the following restrictions:
 *
 * 1. The author is not responsible for the consequences of use of this
 *    software, no matter how awful, even if they arise from flaws in it.
 *
 * 2. The origin of this software must not be misrepresented, either by
 *    explicit claim or by omission.  Since few users ever read sources,
 *    credits must appear in the documentation.
 *
 * 3. Altered versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.  Since few users
 *    ever read sources, credits must appear in the documentation.
 *
 * 4. This notice may not be removed or altered.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/types.h>
#include <sys/param.h>	/* for MAXPATHLEN */
#include <sys/stat.h>
#include <fcntl.h>	/* for open() */
#ifdef RESTORE_TIME
# if (__COHERENT__ >= 0x420)
#  include <sys/utime.h>
# else
#  ifdef USE_UTIMES
#   include <sys/time.h>
#  else
#   include <utime.h>
#  endif
# endif
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>	/* for read() */
#endif
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif

#include <getopt.h>	/* for long options (is this portable?)*/

#include <netinet/in.h>		/* for byte swapping */

#include "file.h"
#include "patchlevel.h"

#ifndef	lint
FILE_RCSID("@(#)$Id: file.c,v 1.59 2001/07/23 00:02:32 christos Exp $")
#endif	/* lint */


#ifdef S_IFLNK
# define USAGE  "Usage: %s [-bciknsvzL] [-f namefile] [-m magicfiles] file...\n"
#else
# define USAGE  "Usage: %s [-bciknsvz] [-f namefile] [-m magicfiles] file...\n"
#endif

#ifndef MAGIC
# define MAGIC "/etc/magic"
#endif

#ifndef MAXPATHLEN
#define	MAXPATHLEN	512
#endif

int 			/* Global command-line options 		*/
	debug = 0, 	/* debugging 				*/
	lflag = 0,	/* follow Symlinks (BSD only) 		*/
	bflag = 0,	/* brief output format	 		*/
	zflag = 0,	/* follow (uncompress) compressed files */
	sflag = 0,	/* read block special files		*/
	iflag = 0,
	nobuffer = 0,   /* Do not buffer stdout */
	kflag = 0;	/* Keep going after the first match	*/

int			/* Misc globals				*/
	nmagic = 0;	/* number of valid magic[]s 		*/

struct  magic *magic;	/* array of magic entries		*/

const char *magicfile = 0;	/* where the magic is		*/
const char *default_magicfile = MAGIC;

char *progname;		/* used throughout 			*/
int lineno;		/* line number in the magic file	*/


static void	unwrap		__P((char *fn));
static void	usage		__P((void));
#if 0
static int	byteconv4	__P((int, int, int));
static short	byteconv2	__P((int, int, int));
#endif

int main __P((int, char *[]));

void help()
{
	puts(
"Usage: file [OPTION]... [FILE]...
Determine file type of FILEs.

  -m, --magic-file LIST      use LIST as a colon-separated list of magic
                               number files
  -z, --uncompress           try to look inside compressed files
  -b, --brief                do not prepend filenames to output lines
  -c, --checking-printout    print the parsed form of the magic file, use in
                               conjunction with -m to debug a new magic file
                               before installing it
  -f, --files-from FILE      read the filenames to be examined from FILE
  -i, --mime                 output mime type strings
  -k, --keep-going           don't stop at the first match
  -L, --dereference          causes symlinks to be followed
  -n, --no-buffer            do not buffer output
  -s, --special-files        treat special (block/char devices) files as
                             ordinary ones
      --help                 display this help and exit
      --version              output version information and exit");
	exit(0);
}

/*
 * main - parse arguments and handle options
 */
int
main(argc, argv)
	int argc;
	char *argv[];
{
	int c, longindex;
	int action = 0, didsomefiles = 0, errflg = 0, ret = 0, app = 0;
	char *mime, *home, *usermagic;
	struct stat sb;
	static struct option long_options[] =
	{
		{"version", 0, 0, 'v'},
		{"help", 0, 0, 0},
		{"brief", 0, 0, 'b'},
		{"checking-printout", 0, 0, 'c'},
		{"debug", 0, 0, 'd'},
		{"files-from", 1, 0, 'f'},
		{"mime", 0, 0, 'i'},
		{"keep-going", 0, 0, 'k'},
#ifdef S_IFLNK
		{"dereference", 0, 0, 'L'},
#endif
		{"magic-file", 1, 0, 'm'},
		{"uncompress", 0, 0, 'z'},
		{"no-buffer", 0, 0, 'n'},
		{"special-files", 0, 0, 's'},
		{"compile", 0, 0, 'C'},
		{0, 0, 0, 0},
	};

#ifdef LC_CTYPE
	setlocale(LC_CTYPE, ""); /* makes islower etc work for other langs */
#endif

	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	magicfile = default_magicfile;
	if ((usermagic = getenv("MAGIC")) != NULL)
		magicfile = usermagic;
	else
		if ((home = getenv("HOME")) != NULL) {
			if ((usermagic = malloc(strlen(home) + 8)) != NULL) {
				(void)strcpy(usermagic, home);
				(void)strcat(usermagic, "/.magic");
				if (stat(usermagic, &sb)<0) 
					free(usermagic);
				else
					magicfile = usermagic;
			}
		}

	while ((c = getopt_long(argc, argv, "bcdf:ikm:nsvzCL", long_options, &longindex)) != EOF)
		switch (c) {
		case 0 :
			if (longindex==1)
				help();
			break;
		case 'b':
			++bflag;
			break;
		case 'c':
			action = CHECK;
			break;
		case 'C':
			action = COMPILE;
			break;
		case 'd':
			++debug;
			break;
		case 'f':
			if (!app) {
				ret = apprentice(magicfile, action);
				if (action)
					exit(ret);
				app = 1;
			}
			unwrap(optarg);
			++didsomefiles;
			break;
		case 'i':
			iflag++;
			if ((mime = malloc(strlen(magicfile) + 6)) != NULL) {
				(void)strcpy(mime, magicfile);
				(void)strcat(mime, ".mime");
				magicfile = mime;
			}
			break;
		case 'k':
			kflag = 1;
			break;
		case 'm':
			magicfile = optarg;
			break;
		case 'n':
			++nobuffer;
			break;
		case 's':
			sflag++;
			break;
		case 'v':
  			(void) fprintf(stdout, "%s-"VERSION"\n", progname);
			(void) fprintf(stdout, "magic data from %s\n",
				       magicfile);
			return 1;
		case 'z':
			zflag++;
			break;
#ifdef S_IFLNK
		case 'L':
			++lflag;
			break;
#endif
		case '?':
		default:
			errflg++;
			break;
		}

	if (errflg) {
		usage();
		fputs("Try `file --help' for more information.\n",stderr);
	}

	if (!app) {
		ret = apprentice(magicfile, action);
		if (action)
			exit(ret);
		app = 1;
	}

	if (optind == argc) {
		if (!didsomefiles) {
			usage();
		}
	}
	else {
		int i, wid, nw;
		for (wid = 0, i = optind; i < argc; i++) {
			nw = strlen(argv[i]);
			if (nw > wid)
				wid = nw;
		}
		for (; optind < argc; optind++)
			process(argv[optind], wid);
	}

	return 0;
}


/*
 * unwrap -- read a file of filenames, do each one.
 */
static void
unwrap(fn)
	char *fn;
{
	char buf[MAXPATHLEN];
	FILE *f;
	int wid = 0, cwid;

	if (strcmp("-", fn) == 0) {
		f = stdin;
		wid = 1;
	} else {
		if ((f = fopen(fn, "r")) == NULL) {
			error("Cannot open `%s' (%m).\n", fn);
			/*NOTREACHED*/
		}

		while (fgets(buf, MAXPATHLEN, f) != NULL) {
			cwid = strlen(buf) - 1;
			if (cwid > wid)
				wid = cwid;
		}

		rewind(f);
	}

	while (fgets(buf, MAXPATHLEN, f) != NULL) {
		buf[strlen(buf)-1] = '\0';
		process(buf, wid);
		if(nobuffer)
			(void) fflush(stdout);
	}

	(void) fclose(f);
}


#if 0
/*
 * byteconv4
 * Input:
 *	from		4 byte quantity to convert
 *	same		whether to perform byte swapping
 *	big_endian	whether we are a big endian host
 */
static int
byteconv4(from, same, big_endian)
	int from;
	int same;
	int big_endian;
{
	if (same)
		return from;
	else if (big_endian) {		/* lsb -> msb conversion on msb */
		union {
			int i;
			char c[4];
		} retval, tmpval;

		tmpval.i = from;
		retval.c[0] = tmpval.c[3];
		retval.c[1] = tmpval.c[2];
		retval.c[2] = tmpval.c[1];
		retval.c[3] = tmpval.c[0];

		return retval.i;
	}
	else
		return ntohl(from);	/* msb -> lsb conversion on lsb */
}

/*
 * byteconv2
 * Same as byteconv4, but for shorts
 */
static short
byteconv2(from, same, big_endian)
	int from;
	int same;
	int big_endian;
{
	if (same)
		return from;
	else if (big_endian) {		/* lsb -> msb conversion on msb */
		union {
			short s;
			char c[2];
		} retval, tmpval;

		tmpval.s = (short) from;
		retval.c[0] = tmpval.c[1];
		retval.c[1] = tmpval.c[0];

		return retval.s;
	}
	else
		return ntohs(from);	/* msb -> lsb conversion on lsb */
}
#endif

/*
 * process - process input file
 */
void
process(inname, wid)
	const char	*inname;
	int wid;
{
	int	fd = 0;
	static  const char stdname[] = "standard input";
	unsigned char	buf[HOWMANY+1];	/* one extra for terminating '\0' */
	struct stat	sb;
	int nbytes = 0;	/* number of bytes read from a datafile */
	char match = '\0';

	if (strcmp("-", inname) == 0) {
		if (fstat(0, &sb)<0) {
			error("cannot fstat `%s' (%m).\n", stdname);
			/*NOTREACHED*/
		}
		inname = stdname;
	}

	if (wid > 0 && !bflag)
	     (void) printf("%s:%*s ", inname, 
			   (int) (wid - strlen(inname)), "");

	if (inname != stdname) {
		/*
		 * first try judging the file based on its filesystem status
		 */
		if (fsmagic(inname, &sb) != 0) {
			putchar('\n');
			return;
		}

		if ((fd = open(inname, O_RDONLY)) < 0) {
			/* We can't open it, but we were able to stat it. */
			if (sb.st_mode & 0002) ckfputs("writeable, ", stdout);
			if (sb.st_mode & 0111) ckfputs("executable, ", stdout);
			ckfprintf(stdout, "can't read `%s' (%s).\n",
			    inname, strerror(errno));
			return;
		}
	}


	/*
	 * try looking at the first HOWMANY bytes
	 */
	if ((nbytes = read(fd, (char *)buf, HOWMANY)) == -1) {
		error("read failed (%m).\n");
		/*NOTREACHED*/
	}

	if (nbytes == 0)
		ckfputs(iflag ? "application/x-empty" : "empty", stdout);
	else {
		buf[nbytes++] = '\0';	/* null-terminate it */
		match = tryit(buf, nbytes, zflag);
	}

#ifdef BUILTIN_ELF
	if (match == 's' && nbytes > 5) {
		/*
		 * We matched something in the file, so this *might*
		 * be an ELF file, and the file is at least 5 bytes long,
		 * so if it's an ELF file it has at least one byte
		 * past the ELF magic number - try extracting information
		 * from the ELF headers that can't easily be extracted
		 * with rules in the magic file.
		 */
		tryelf(fd, buf, nbytes);
	}
#endif

	if (inname != stdname) {
#ifdef RESTORE_TIME
		/*
		 * Try to restore access, modification times if read it.
		 * This is really *bad* because it will modify the status
		 * time of the file... And of course this will affect
		 * backup programs
		 */
# ifdef USE_UTIMES
		struct timeval  utsbuf[2];
		utsbuf[0].tv_sec = sb.st_atime;
		utsbuf[1].tv_sec = sb.st_mtime;

		(void) utimes(inname, utsbuf); /* don't care if loses */
# else
		struct utimbuf  utbuf;

		utbuf.actime = sb.st_atime;
		utbuf.modtime = sb.st_mtime;
		(void) utime(inname, &utbuf); /* don't care if loses */
# endif
#endif
		(void) close(fd);
	}
	(void) putchar('\n');
}


int
tryit(buf, nb, zfl)
	unsigned char *buf;
	int nb, zfl;
{
	/* try compression stuff */
	if (zfl && zmagic(buf, nb))
		return 'z';

	/* try tests in /etc/magic (or surrogate magic file) */
	if (softmagic(buf, nb))
		return 's';

	/* try known keywords, check whether it is ASCII */
	if (ascmagic(buf, nb))
		return 'a';

	/* abandon hope, all ye who remain here */
	ckfputs("data", stdout);
		return '\0';
}

static void
usage()
{
	(void)fprintf(stderr, USAGE, progname);
	(void)fprintf(stderr, "Usage: %s -C [-m magic]\n", progname);
	exit(1);
}
